contributors/design-proposals/defer-containers.md

```diff
@@ -0,0 +1,248 @@
+# Defer Containers
+
+dhilip.kumar.s@huawei.com
+
+March 2017
+
+## Prerequisite
+This is a continuation of [initContainers](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/container-init.md)
+
+## Motivation
+Since the introduction to Statefulset and Daemonset PODs that consist of stateful containers are now fully supported by
+Kubernetes.  Such stateful pods benefit from constructor and destructor semantics. InitContainers provide constructor
+semantics. The proposed DeferContainers will provide destructor semantics for such stateful PODs
+
+## Abstract
+This introduces the concept of ‘deferContainers’, inspired by golang’s keyword ‘defer’. This will allow one to define a
+set of Containers to be executed at the time of POD termination. The defined containers will be executed sequentially and
+in the specified order. It will bring destructor() capability to Pods.
+
+## Use Cases
+
+### Cleanup:
```

Comment:
@smarterclayton I have added some design thoughts on how we handle time boundness for deferContainers,  this was something i wanted to bring up in the SIG-APPs meeting yesterday.   PTAL. 

Also this particular discussion is about deleting the POD without running deferContainers or preStopHooks,  an equivalent of `kill -9 PID`.  Could you please elaborate why should we come up with a different signal for this?